Add upgrade documentation for Destinations V2 (#27334)
airbytehq/airbyte: docs/release_notes/upgrading_to_destinations_v2.md

# Upgrading to Destinations V2

## What is Destinations V2?

> At launch, Airbyte Destinations V2 provides:

Reviewer: Do you think we need to do more 'selling' here?

Author: I think the blog post should do the selling, and the docs should be more straight to the point. It would be worth linking to the blog post when that is ready.

|-----------------------------------------------	|-----------------	|-------------------------------------------------------------------------------------------------------------	|----------------------	|---------------------	|
| Successful typing and de-duping ⟶             	| xxx-xxx-xxx     	| {   id: 1,   first_name: “sarah”,   age: 39,   address: {     city: “San Francisco”,     zip: “94131”   } } 	| 2022-01-01 12:00:001 	| 2022-01-01 12:00:00﻿ 	|
| Failed typing that didn’t break other rows ⟶  	| yyy-yyy-yyy     	| {   id: 2,   first_name: “evan”,   age: “fish”,   address: {     city: “Menlo Park”,     zip: “94002”   } } 	| 2022-01-01 12:00:001 	| 2022-01-01 12:00:00﻿ 	|
| Not-yet-typed ⟶                               	| zzz-zzz-zzz     	| {   id: 3,   first_name: “edward”,   age: 35,   address: {     city: “Sunnyvale”,     zip: “94003”   } }    	| NULL                 	| 2022-01-01 13:00:00﻿ 	|
_____________________________

## Deprecating Legacy Normalization

> 

Reviewer: I think you need to start with a notion that "the upgrade to destinations v2 is handled by moving your syncs to use v2 of the snowflake, bigquery, or redshift airbyte destinations.  You will be be able to use the existing v1 destinations until October 1.

Existing normalization options, both `Raw data (JSON)` and `Normalized tabular data` are now deprecated, and will be unsupported starting Oct 1, 2023.

![Legacy Normalization](./assets/airbyte_legacy_normalization.png)

As a Cloud user, existing syncs will be paused on **Oct 1, 2023**. Follow the upgrade steps below to benefit from Destinations V2.

As an Open Source user, you may choose to upgrade at your convenience. However, destination connector versions prior to Destinations V2 will no longer be supported as of **Oct 1, 2023**.

See [here](to-do-doc) to learn more about Airbyte's breaking change rollout requirements.

### Breakdown of Breaking Changes

> Breaking changes will be most notable to users actively syncing data from API sources such as Facebook Marketing, HubSpot, Stripe, Amazon Ads, TikTok Marketing and more. The following table details the delivered data modified by Destinations V2:

Author: Removing, probably not a helpful sentence / as you pointed out, might lead to more questions.

| Current Normalization Setting   	 | Source Type                           	| Impacted Data (Breaking Changes)                         	                     |
|-----------------------------------|---------------------------------------	|--------------------------------------------------------------------------------|
| Raw JSON                	         | All                                   	| `_airbyte` metadata columns, raw table location                              	 |
| Normalized tabular data 	         | API Source                            	| Unnested tables, `_airbyte` metadata columns, SCD tables 	                     |
| Normalized tabular data 	         | Tabular Source (database, file, etc.) 	| `_airbyte` metadata columns, SCD tables                  	                     |

> For `destination-bigquery`, we are now loading `JSON` blobs as type `JSON` in BigQuery (introduced last [year](https://cloud.google.com/blog/products/data-analytics/bigquery-now-natively-supports-semi-structured-data)), instead of type `string`.

Reviewer: ```suggestion
Whenever possible, we've taken this opportunity to use the best data type for storing JSON for your querying connivence.  For example, `destination-bigquery` now loads `JSON` blobs as type `JSON` in BigQuery (introduced last [year](https://cloud.google.com/blog/products/data-analytics/bigquery-now-natively-supports-semi-structured-data)), instead of type `string`.
```

## Quick Start to Upgrading

The quickest path to upgrading is to click upgrade on any out-of-date connection in the UI:

![Upgrade Path](./assets/airbyte_destinations_v2_upgrade_prompt.png)

After upgrading the out-of-date destination to a [Destinations V2 compatible version](#destinations-v2-effective-versions), the following will occur at the next sync **for each connection** sending data to the updated destination:
1. Existing raw tables replicated to this destination will be copied to a new `airbyte` schema. 
2. The copied raw table will be updated to the new Destinations V2 format.
3. The new raw tables will be updated with any new data since the last sync.
4. The updated raw tables will be typed and de-duplicated according to the Destinations V2 format.
5. Once typing and de-duplication has completed successfully, your previous final table will be replaced with the updated data.

> Pre-existing raw tables, SCD tables and "unnested" tables will always be left untouched. You can delete these at your convenience.

Reviewer: ```suggestion
Pre-existing raw tables, SCD tables and "unnested" tables will always be left untouched. You can delete these at your convenience.  Note that they will no longer be updated by your syncs.
```

Versions are tied to the destination. When you update the destination, **all connections tied to that destination will be sending data in the Destinations V2 format**. For upgrade paths that will minimize disruption to existing dashboards, see:
* [Upgrading Connections One by One with Dual-Writing](#upgrading-connections-one-by-one-with-dual-writing)
* [Testing Destinations V2 on a Single Connection](#testing-destinations-v2-for-a-single-connection)
* [Upgrading as a User of Raw Tables](#upgrading-as-a-user-of-raw-tables)
* [(OSS Only) Rolling back to Legacy Normalization](#oss-only-rolling-back-to-legacy-normalization)
 
## Advanced Upgrade Paths

### Upgrading Connections One by One with Dual-Writing

Dual writing is a method employed during upgrades where new incoming data is written simultaneously to both the old and new systems, facilitating a smooth transition between versions. We recommend this approach for connections where you are especially worried about breaking changes or downtime in downstream systems.

#### Steps to Follow for All Sync Modes

> 1. **[Open Source]** Update the default destination version for your workspace to a [Destinations V2 compatible version](#destinations-v2-effective-versions). This sets the default version for any newly created destination.

Reviewer: ```suggestion
1. **[Open Source]** Update the default destination version for your workspace to a [Destinations V2 compatible version](#destinations-v2-effective-versions). This sets the default version for any newly created destination.  Your existing syncs will remain on the previous version 
```

* assuming the OSS deployment has opted into breaking change management.  @erohmensing is working though this now

![Upgrade your default destination version](assets/airbyte_version_upgrade.png)

2. Create and configure a new destination connecting to the same database as your existing destination except for `Default Schema`, which you should update to a new value to avoid collisions.

![Create a new destination](assets/airbyte_dual_destinations.png)

3. Create a new connection leveraging your existing source and the newly created destination. Match the settings of your pre-existing connection.
4. If the streams you are looking to replicate are in **full refresh** mode, enabling the connection will now provide a parallel copy of the data in the updated format for testing. If any of the streams in the connection are in an **incremental** sync mode, follow the steps below before enabling the connection.

> #### Additional Steps for Incremental Sync Modes

Author: Add exceptions for Postgres CDC

Author: MySQL CDC may have unknowns.

Reviewer: > Does copying the state work (for CDC)?

No. For postgres, this is tricky.  A new CDC Replication Slot will need to be created as well, and the sync will *have* to start from scratch

These steps allow you to dual-write for connections incrementally syncing data without re-syncing historical data you've already replicated:

1. Copy the raw data you've already replicated to the new schema being used by your newly created connection. You need to do this for every stream in the connection with an incremental sync mode. Sample SQL you can run in your data warehouse:

```mysql
CREATE TABLE {new_schema}.raw_{stream_name} AS 
SELECT * 
FROM {old_schema}.raw_{stream_name};

> ```

Reviewer: Evan TODO:
* We'll need to make the new table first with the right columns and types.  This query will differ per destination
* We will load each column by name to handle the rename (_emitted_at -> _extracted_at)

2. Go to your newly created connection, and navigate to the `Settings` tab.

> 3. Open the `Advanced` settings to see the connection state (which manages incremental syncs). Select the existing connection you are duplicating, then click `Inherit State`. This will ensure historical data is not replicated again.

Reviewer: Neat.

* When you upgrade to a [Destinations V2 compatible version](#destinations-v2-effective-versions) of your destination, we will never alter your existing raw data. Although existing downstream dashboards will go stale, they will never be broken.
* You can dual write by following the [steps above](#upgrading-connections-one-by-one-with-dual-writing) and copying your raw data to the schema of your newly created connection.

We may make further changes to raw tables in the future, as these tables are intended to be a staging ground for Airbyte to optimize the performance of your syncs. We cannot guarantee the same level of stability as for final tables in your destination schema.

### (OSS Only) Rolling back to Legacy Normalization

We have published a [rollback version for each destination](#destinations-v2-compatible-versions) that will re-create the final tables with normalization using raw tables in the new format if they are available, and otherwise default to pre-existing raw tables used by legacy normalization.

## Destinations V2 Compatible Versions

For each destination connector, Destinations V2 is effective as of the following versions:

> | Destination Connector 	| Safe Rollback Version 	| Destinations V2 Compatible 	 |

Reviewer: We'll be bumping the first (major) number for 1s1t.

|------------	|----------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Postgres   	| [Upgrade connection in place](#quick-start-to-upgrading)	      | You can optionally dual write, but this requires resyncing historical data from the source. You must create a new Postgres source with a different replication slot than your existing source to preserve the integrity of your existing connection.             	 |
| MySQL      	| [All above upgrade paths supported](#advanced-upgrade-paths) 	 | You can upgrade the connection in place, or dual write. When dual writing, Airbyte can leverage the state of an existing, active connection to ensure historical data is not re-replicated from MySQL. 	                                                           |

> | SQL Server 	| [All above upgrade paths supported](#advanced-upgrade-paths) 	 | You can upgrade the connection in place, or dual write. When dual writing, Airbyte can leverage the state of an existing, active connection to ensure historical data is not re-replicated from SQL Server. 	                                                      |

Author: This needs to be confirmed.
